test/test_build_stream.ts

```diff
@@ -0,0 +1,191 @@
+/**
+ * @license
+ * Copyright 2018 Balena
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Bluebird from 'bluebird';
+import { Stream, Readable, Writable, Duplex } from 'stream';
+
+import Builder from '../src/index';
+import { sampleDaemonOutputGenerator, sampleDaemonStreamGenerator }
+	from './test-files/sample_daemon_output';
+
+const assert = require('assert').strict;
+
+/**
+ * This test asserts that the createBuildStream() method writes the expected
+ * data to the stream it returns, also prints how it long it took to write the
+ * tar stream to a mocked docker daemon and how long it took to read and parse
+ * the stream/status data from the mocked daemon.  (Note that there are
+ * additional tests in tests.ts that cover other additional functionality such
+ * as calling the success/failure hooks as expected.)
+ *
+ * The test is run only once so obviously does provide accurate time figures at
+ * all, but importantly it allows developers to play with the implementation of
+ * createBuildStream, and then run the test multiple times for the old and new
+ * implementations to compare the performance impact.
+ * 
+ * As is, it takes under 50ms to run in a devenv VirtualBox VM on my laptop.
+ */
+describe('createBuildStream performance', function() {
+	this.timeout(5000);
+
+	it('should be fast', () => {
+		const streamer = sampleDaemonStreamGenerator();
+		const mockBuilder = new MockBuilder();
+		let startTime: number;
+		let buildStream: Duplex;
+		const buildStreamPromise = new Bluebird((resolve, reject) => {
+			
+			buildStream = Builder.prototype.createBuildStream.call(mockBuilder, {})
+			.on('error', (error: Error) => reject(error))
+			.on('end', () => {
+				console.log('createBuildStream performance test: write time (tar stream): '
+				+ `${mockBuilder.docker.tarStreamMilliseconds} milliseconds`);
+
+				console.log('createBuildStream performance test: read time (JSON stream): '
+					+ `${Date.now() - startTime} milliseconds`);
```

Comment:
I wrote this test because I wanted to have an idea of the performance impact of replacing `event-stream` `es.through()` with node's `Stream.PassThrough` in the implementation of `createBuildStream()`. I've found that `event-stream` is faster, and I guess it's because it passes buffer/string references down the pipe instead of copying the data:

Tar Stream Size | es.through() | Stream.PassThrough()
--- | --- | ---
1MB | ~20ms | ~35ms
10MB | ~70ms | ~100ms
100MB | ~190ms | ~250ms
1GB | ~1300ms | ~1650ms

The numbers were taken by running this test on the devenv (VirtualBox VM on my laptop). The relative difference is more relevant than the absolute numbers, but it's good to have an idea of the magnitude (milliseconds).

As I understand it (please challenge it if it doesn't sound right), these streams are used to read/write data to a local socket between the builder node.js application and a local docker daemon. A tar stream is piped to the daemon, and a JSON stream (status messages) is read from the daemon. I read recently on Flowdock some messages from Petros where he shared some builder stats that showed build times from a few minutes to over 30 minutes, with the average around 30 minutes if I recall correctly (I can't find it in Flowdock anymore). If this is correct _(again, please challenge it),_ then the performance difference in question is a matter of 15 to 350 milliseconds in builds that take several minutes, and therefore a **negligible difference.** (I don't know the average tar stream length in bytes but I guess it's much less than 1GB.)

Does this reasoning sound correct?

The only reason to replace `es.through` with `Stream.PassThrough` is to remove the dependency on `event-stream` following the `flatmap` security incident, but I have no problems at all with keeping `event-stream` and reducing the pull request to merely add the `if` test that fixes the "false positive build" bug. (I might want to keep this test file because it was handy to iteratively compare changes to the implementation of `createBuildStream()`.)

Reply:
@CameronDiver, @Page-, given this comment, does it still make sense to remove the `event-stream` dependency from `resin-docker-build`?

Reply:
@zvin, thanks for the review once again. Following a chat with @CameronDiver, we decided that there wasn't enough justification to remove the dependency on `event-stream`, so I've reverted the original implementation and simply added a `!errored` test. I've kept the additional test code, though.